In [1]:
import pandas as pd

In [14]:
df = pd.read_csv("/home/pdutta/Data/GDC/test.tsv", sep="\t", header=None, on_bad_lines='skip')

In [15]:
df

0    1           2       3   4   \
0         UNC14-SN744_119:5:1106:11512:33180  163        chr1   44171   3   
1         UNC14-SN744_119:5:2201:20042:92204  385        chr1  181527   3   
2         UNC14-SN744_119:5:2205:6445:126153  163        chr1  181584   3   
3        UNC14-SN744_119:5:2107:12291:125745   65        chr1  184136   3   
4          UNC14-SN744_119:5:1107:2174:82112  129        chr1  184150   3   
...                                      ...  ...         ...     ...  ..   
1601619  UNC14-SN744_119:5:2206:14179:121441  385      HPV162    5877   3   
1601620  UNC14-SN744_119:5:1104:21254:147738  433      HPV180    3285   3   
1601621    UNC14-SN744_119:5:2108:12674:5432  385    HPV-mCG2    1443   3   
1601622   UNC14-SN744_119:5:1105:5584:193888  369  HPV-mFi864     226   3   
1601623   UNC14-SN744_119:5:2102:9391:196933  433    HPV-mKN2    6935   3   

                5      6          7       8   \
0           35S15M      =     632115  587994   
1         8S21M21S  chr19    2101317       0   
2           33S17M      =     631598  450064   
3              50M   chr2  113599690       0   
4              50M   chr2  113599633       0   
...            ...    ...        ...     ...   
1601619  10S19M21S  chr19   46775234       0   
1601620     31S19M   chr4    2928464       0   
1601621   1S23M26S  chr13   44239083       0   
1601622   25S17M8S   chr1  153991189       0   
1601623  22S18M10S   chrM       5342       0   

                                                        9   \
0        CAAAAAAAAAAAAAAAAAAAAAAAAAAAAATAAACAATAAAAAAAA...   
1        CTCGGCTTGGGTTTGGGGGGGGCGTGTGTGGTCCGCATCCCCTGCT...   
2        GAGGGGCGTTTGGTATTGGGTTATGGCAGGGGGGGGGGGGGGGGGG...   
3        CAGGACCCAGGCACAGGCATTAGTGCCCGTTGGAGAAAACGGGAAT...   
4        AGGCATTAGTGCCCGTTGGAGAAAACGGGAATCCCGAAGAAATGGT...   
...                                                    ...   
1601619  AGAAAACAGCAAGATGGGGACATGAGTGAGATCTGGGATGCCTGGC...   
1601620  CAGCCACCGGGGCTGGGTCTGGGGCTGGCTCAGTCCAAAAAGAAGA...   
1601621  GGCAAATGATTCTTTCTGATTTTTTTCTCTGCCAGAACACTAGGAT...   
1601622  GGAGAGGGATGAAGGAGATCCTTTGAGCGGTGGTGATTTTATAGCA...   
1601623  GATGATGAGGCTATTGTTTTTTTGTGAATTCTAACGTTGTTAGATA...   

                                                        10      11      12  \
0        +1+=BD:A@D####################################...  NH:i:2  HI:i:1   
1        CCCFFFFF<FH+CF@FIJ)5:B6@790?2?B8@C67167>BD?<?>...  NH:i:2  HI:i:2   
2        BCCAFFFFHHHDH3CBFGIGDHIGEHI?;<:FB#############...  NH:i:2  HI:i:1   
3        CCCFFFFFHHHHHJJJJJJJJJJHIIIJJHIJJHIGIJIJJJJHJJ...  NH:i:2  HI:i:1   
4        @CCFFFFFFFDHHJIIJJJIIJIJGHIIGIJIJJJJIIIGEGGIJF...  NH:i:2  HI:i:2   
...                                                    ...     ...     ...   
1601619  CCCFFFFFHHHHHJJJJIJJGIJIIG?FHGGGHJJIJEHGGHGIGH...  NH:i:2  HI:i:2   
1601620  B2GBF;;GFD;@DB0))?;?::);D>HHCG9HGBDDGDFDHHDFFF...  NH:i:2  HI:i:2   
1601621  CCCFFFFFHHHHHJJJIJIJJJJJJJJJJJJJJJIIJJJJIIJJJJ...  NH:i:2  HI:i:2   
1601622  JJJJJJJJJJJJJJJJJJJIJJJJJJJJJJJJJJHJJHHHHGFFFF...  NH:i:2  HI:i:1   
1601623  JJJJJJJJJIJJJJJJJJIJJJJJJJJJJJJJJHJJJHHHHHFFFF...  NH:i:2  HI:i:2   

              13      14      15                                         16  
0        AS:i:60  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1        AS:i:18  nM:i:1  NM:i:1  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
2        AS:i:58  nM:i:2  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
3        AS:i:49  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
4        AS:i:49  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
...          ...     ...     ...                                        ...  
1601619  AS:i:18  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1601620  AS:i:18  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1601621  AS:i:22  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1601622  AS:i:16  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  


In [16]:
df[1].unique()

array([163, 385,  65, 129, 113, 177, 161,  97, 145,  81, 417,  83, 401,
       337,  99, 433, 369, 147, 353, 321])

In [17]:
df[df[0].duplicated(keep=False)]

0    1           2       3   4   \
0         UNC14-SN744_119:5:1106:11512:33180  163        chr1   44171   3   
1         UNC14-SN744_119:5:2201:20042:92204  385        chr1  181527   3   
2         UNC14-SN744_119:5:2205:6445:126153  163        chr1  181584   3   
3        UNC14-SN744_119:5:2107:12291:125745   65        chr1  184136   3   
4          UNC14-SN744_119:5:1107:2174:82112  129        chr1  184150   3   
...                                      ...  ...         ...     ...  ..   
1601619  UNC14-SN744_119:5:2206:14179:121441  385      HPV162    5877   3   
1601620  UNC14-SN744_119:5:1104:21254:147738  433      HPV180    3285   3   
1601621    UNC14-SN744_119:5:2108:12674:5432  385    HPV-mCG2    1443   3   
1601622   UNC14-SN744_119:5:1105:5584:193888  369  HPV-mFi864     226   3   
1601623   UNC14-SN744_119:5:2102:9391:196933  433    HPV-mKN2    6935   3   

                5      6          7       8   \
0           35S15M      =     632115  587994   
1         8S21M21S  chr19    2101317       0   
2           33S17M      =     631598  450064   
3              50M   chr2  113599690       0   
4              50M   chr2  113599633       0   
...            ...    ...        ...     ...   
1601619  10S19M21S  chr19   46775234       0   
1601620     31S19M   chr4    2928464       0   
1601621   1S23M26S  chr13   44239083       0   
1601622   25S17M8S   chr1  153991189       0   
1601623  22S18M10S   chrM       5342       0   

                                                        9   \
0        CAAAAAAAAAAAAAAAAAAAAAAAAAAAAATAAACAATAAAAAAAA...   
1        CTCGGCTTGGGTTTGGGGGGGGCGTGTGTGGTCCGCATCCCCTGCT...   
2        GAGGGGCGTTTGGTATTGGGTTATGGCAGGGGGGGGGGGGGGGGGG...   
3        CAGGACCCAGGCACAGGCATTAGTGCCCGTTGGAGAAAACGGGAAT...   
4        AGGCATTAGTGCCCGTTGGAGAAAACGGGAATCCCGAAGAAATGGT...   
...                                                    ...   
1601619  AGAAAACAGCAAGATGGGGACATGAGTGAGATCTGGGATGCCTGGC...   
1601620  CAGCCACCGGGGCTGGGTCTGGGGCTGGCTCAGTCCAAAAAGAAGA...   
1601621  GGCAAATGATTCTTTCTGATTTTTTTCTCTGCCAGAACACTAGGAT...   
1601622  GGAGAGGGATGAAGGAGATCCTTTGAGCGGTGGTGATTTTATAGCA...   
1601623  GATGATGAGGCTATTGTTTTTTTGTGAATTCTAACGTTGTTAGATA...   

                                                        10      11      12  \
0        +1+=BD:A@D####################################...  NH:i:2  HI:i:1   
1        CCCFFFFF<FH+CF@FIJ)5:B6@790?2?B8@C67167>BD?<?>...  NH:i:2  HI:i:2   
2        BCCAFFFFHHHDH3CBFGIGDHIGEHI?;<:FB#############...  NH:i:2  HI:i:1   
3        CCCFFFFFHHHHHJJJJJJJJJJHIIIJJHIJJHIGIJIJJJJHJJ...  NH:i:2  HI:i:1   
4        @CCFFFFFFFDHHJIIJJJIIJIJGHIIGIJIJJJJIIIGEGGIJF...  NH:i:2  HI:i:2   
...                                                    ...     ...     ...   
1601619  CCCFFFFFHHHHHJJJJIJJGIJIIG?FHGGGHJJIJEHGGHGIGH...  NH:i:2  HI:i:2   
1601620  B2GBF;;GFD;@DB0))?;?::);D>HHCG9HGBDDGDFDHHDFFF...  NH:i:2  HI:i:2   
1601621  CCCFFFFFHHHHHJJJIJIJJJJJJJJJJJJJJJIIJJJJIIJJJJ...  NH:i:2  HI:i:2   
1601622  JJJJJJJJJJJJJJJJJJJIJJJJJJJJJJJJJJHJJHHHHGFFFF...  NH:i:2  HI:i:1   
1601623  JJJJJJJJJIJJJJJJJJIJJJJJJJJJJJJJJHJJJHHHHHFFFF...  NH:i:2  HI:i:2   

              13      14      15                                         16  
0        AS:i:60  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1        AS:i:18  nM:i:1  NM:i:1  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
2        AS:i:58  nM:i:2  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
3        AS:i:49  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
4        AS:i:49  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
...          ...     ...     ...                                        ...  
1601619  AS:i:18  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1601620  AS:i:18  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1601621  AS:i:22  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1601622  AS:i:16  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  


## FLAG meaning

Let's decode each of these SAM flag values to understand their meanings:

1. **163**: 
    - Paired in sequencing (`1`).
    - Properly paired (`2`).
    - Mate is mapped to reverse strand (`32`).
    - Second in pair (`128`).

2. **385**:
    - Paired in sequencing (`1`).
    - Mate is unmapped (`8`).
    - Read is not primary alignment (`256`).
    - First in pair (`64`).
  
3. **65**: 
    - Paired in sequencing (`1`).
    - First in pair (`64`).

4. **129**:
    - Paired in sequencing (`1`).
    - Second in pair (`128`).

5. **113**: 
    - Paired in sequencing (`1`).
    - Mapped to reverse strand (`16`).
    - First in pair (`64`).
    - Mate is unmapped (`32`).

6. **177**: 
    - Paired in sequencing (`1`).
    - Mapped to reverse strand (`16`).
    - First in pair (`64`).
    - Mate is mapped to reverse strand (`32`).
    - Supplementary alignment (`128`).

7. **161**: 
    - Paired in sequencing (`1`).
    - Second in pair (`128`).
    - Mate is unmapped (`32`).

8. **97**: 
    - Paired in sequencing (`1`).
    - Mapped to reverse strand (`16`).
    - First in pair (`64`).
    - Mate is unmapped (`16`).

9. **145**: 
    - Paired in sequencing (`1`).
    - Second in pair (`128`).
    - Mate is mapped to reverse strand (`16`).

10. **81**: 
    - Paired in sequencing (`1`).
    - First in pair (`64`).
    - Read is not primary alignment (`16`).

11. **417**: 
    - Paired in sequencing (`1`).
    - First in pair (`64`).
    - Read is not primary alignment (`256`).
    - Supplementary alignment (`128`).

12. **83**: 
    - Paired in sequencing (`1`).
    - Properly paired (`2`).
    - Mapped to reverse strand (`16`).
    - First in pair (`64`).

13. **401**: 
    - Paired in sequencing (`1`).
    - Second in pair (`128`).
    - Read is not primary alignment (`256`).
    - Supplementary alignment (`16`).

14. **337**: 
    - Paired in sequencing (`1`).
    - Mapped to reverse strand (`16`).
    - First in pair (`64`).
    - Mate is mapped to reverse strand (`32`).
    - Read is not primary alignment (`256`).

15. **99**: 
    - Paired in sequencing (`1`).
    - Properly paired (`2`).
    - First in pair (`64`).
    - Mate is mapped to reverse strand (`32`).

16. **433**: 
    - Paired in sequencing (`1`).
    - First in pair (`64`).
    - Supplementary alignment (`128`).
    - Read is not primary alignment (`256`).

17. **369**: 
    - Paired in sequencing (`1`).
    - Unmapped (`8`).
    - First in pair (`64`).
    - Read is not primary alignment (`256`).

18. **147**: 
    - Paired in sequencing (`1`).
    - Properly paired (`2`).
    - Second in pair (`128`).
    - Mate is mapped to reverse strand (`16`).

19. **353**: 
    - Paired in sequencing (`1`).
    - First in pair (`64`).
    - Read is not primary alignment (`256`).
    - Mate is mapped to reverse strand (`32`).

20. **321**: 
    - Paired in sequencing (`1`).
    - Second in pair (`128`).
    - Read is not primary alignment (`256`).
    - Mate is mapped to reverse strand (`32`).



In [18]:
df[df[0]=='UNC14-SN744_119:5:1106:11512:33180']

0    1      2        3   4       5   \
0        UNC14-SN744_119:5:1106:11512:33180  163   chr1    44171   3  35S15M   
11475    UNC14-SN744_119:5:1106:11512:33180   83   chr1   632115   3     50M   
1124630  UNC14-SN744_119:5:1106:11512:33180  417  chr16  4998737   3  35M15S   

           6       7       8   \
0           =  632115  587994   
11475       =   44171 -587994   
1124630  chr1  632115       0   

                                                        9   \
0        CAAAAAAAAAAAAAAAAAAAAAAAAAAAAATAAACAATAAAAAAAA...   
11475    TTTCACCGTAGGTGGCCTGACTGGCATTGTATTAGCAAACTCATCA...   
1124630  CAAAAAAAAAAAAAAAAAAAAAAAAAAAAATAAACAATAAAAAAAA...   

                                                        10      11      12  \
0        +1+=BD:A@D####################################...  NH:i:2  HI:i:1   
11475    CGEIHGIIIIIIHEGIGIIIIIIIIHHHGIHFBIIIIHFFDBFFDD...  NH:i:2  HI:i:1   
1124630  +1+=BD:A@D####################################...  NH:i:2  HI:i:2   

              13      14      15                                         16  
0        AS:i:60  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
11475    AS:i:60  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5  
1124630  AS:i:34  nM:i:0  NM:i:0  RG:Z:110525_UNC14-SN744_0119_BC03HKABXX.5

In [1]:
import pybedtools

In [2]:
bam_file = pybedtools.BedTool("/home/pdutta/Data/GDC/RAMANA/data/BAM/3266ea00-0b4c-4924-bdbe-cc28cca1bde5/5899ee97-f282-4f7d-b6f3-26d3cc78606b.rna_seq.chimeric.gdc_realn.bam")
bed_file = pybedtools.BedTool("/home/pdutta/Data/Splicesites_data/Unique_Acceptor_RAW_Combined_BED.tsv")

In [3]:
intersected = bam_file.intersect(bed_file, u=True)

NotImplementedError: "intersectBed" does not appear to be installed or on the path, so this method is disabled.  Please install a more recent version of BEDTools and re-import to use this method.

In [2]:
import os

file_path = "/Data/GDC/RAMANA/data/BAM/tgfrybgtyiuoj.BAM"

# Split the path to get directory and filename separately
dir_path, filename = os.path.split(file_path)
print(dir_path, filename)
# Extract filename without extension
base_filename = os.path.splitext(filename)[0]
print(base_filename)
# Replace BAM with VCF in directory path
vcf_dir = dir_path.replace("/BAM/", "/VCF/")
print(vcf_dir)
# Construct the new VCF path
vcf_path = os.path.join(vcf_dir, f"{base_filename}.vcf")

print(vcf_path)  # Should print: /Data/GDC/R


/Data/GDC/RAMANA/data/BAM tgfrybgtyiuoj.BAM
tgfrybgtyiuoj
/Data/GDC/RAMANA/data/BAM
/Data/GDC/RAMANA/data/BAM/tgfrybgtyiuoj.vcf
